src/main/java/dk/bec/polonez/reservationsystem/service/OfferService.java

```diff
@@ -47,23 +50,27 @@ public OfferDto getById(long id) {
     }
 
     public OfferDto addOffer(CreateOfferDto offerDto) {
-        if (authService.isPoLoggedIn() || authService.isAdminLoggedIn()) {
+        User currentUser = authService.getCurrentUser();
+        Role currentUserRole = currentUser.getRole();
+        if (currentUserRole.hasOfferCreatePrivilege()) {
             Offer offer = modelMapper.map(offerDto, Offer.class);
-            offer.setOwner(authService.getCurrentUser());
+            offer.setOwner(currentUser);
             Offer savedOffer = offerRepository.save(offer);
             return modelMapper.map(savedOffer, OfferDto.class);
         } else {
-            throw new AuthenticationException(HttpStatus.FORBIDDEN, "Access denied, log as Place Owner or System Admin");
+            throw new NoAccessToOfferOperationException("Access denied, log as Place Owner or System Admin");
         }
     }
 
     public OfferDto updateOffer(long id, CreateOfferDto offerDto) {
         Optional<Offer> optionalOffer = offerRepository.findById(id);
-        if (authService.isPoLoggedIn() || authService.isAdminLoggedIn()) {
-            Offer offer = optionalOffer.orElseThrow(
-                    () -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Offer with id=" + id + " not found"));
+        User currentUser = authService.getCurrentUser();
+        Role currentUserRole = currentUser.getRole();
+        Offer offer = optionalOffer.orElseThrow(
+                () -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Offer with id=" + id + " not found"));
```

Comment:
**warn** I can see that business services are still coupled to implementation details of HTTP protocol in some places. Custom exception should be thrown instead (+ custom exception mapper should be added).

Reply:
They will be edited as I did in addOffer method, I wanted to make commit with not finished service just to show my colleagues a way how to map exceptions.